### DISCUSSION TITLE
Digital Microphone

### DISCUSSION THREAD

**Discussion by CaiusX** [mcguirepr89/BirdNET-Pi#230] [Show and tell]
Hi @nilspupils, @dd4WH and the rest of the gang

I have rigged up two RPi4's recording next to each other, the one my main birds.naturestation.net public system using the Boya-M3 USB mic and the second using the [Seeed 2 Microphone HAT system](https://www.seeedstudio.com/ReSpeaker-2-Mics-Pi-HAT.html)

![image](https://user-images.githubusercontent.com/25373276/166101005-15b5644b-417a-46d4-9cae-d98c180913e7.png)

Location is sub-optimal and I haven't fiddled too much with the ALSA settings for each (same as in https://github.com/mcguirepr89/BirdNET-Pi/discussions/39?sort=new#discussioncomment-2663983).

The SEED needs its [drivers ](https://github.com/respeaker/seeed-voicecard)to be loaded and installed (NOTE USE THE 64bit installer install_arm64.sh).

BirdNET-Pi needs to manually set the card since it isn't recognised out of the box.

In the terminal, get the card number with `arecord -l`

![image](https://user-images.githubusercontent.com/25373276/166101335-e5401ce9-b47c-4021-9494-8b7859554ea1.png)

Card 3 in this case - and put this into the birdnetpi.config - 

![image](https://user-images.githubusercontent.com/25373276/166101454-25fba48c-8f95-42a7-8bfa-76d620445345.png)

Restart all services and away you go.

Sooooooo.... what are the results like?

I'll let the reader make their own minds up (note that neither of my installations are hi-fi, all I'm targeting is workable CHEAP systems)-

Both systems detected this BronZe Mannikin -

![image](https://user-images.githubusercontent.com/25373276/166101694-c6eef5d2-1858-4e5d-8e31-32c7d796643f.png)

The two sound files are:

Boya:

https://user-images.githubusercontent.com/25373276/166101942-c6a5c7b3-78db-47b6-88d1-4600d0188fa3.mp4

SEED

https://user-images.githubusercontent.com/25373276/166104371-3d51ebf7-ef87-4c0e-a771-162a70270d63.mp4

The Boya USB clearly has more "definition" across the whole frequency spectrum, is louder and less noisy. Interestingly the Seeed digital gave a higher confidence for the same sample.

Todays readings for both systems to now are:

![image](https://user-images.githubusercontent.com/25373276/166102230-10f285e7-cda9-4b62-a36e-cb07f8ecf742.png)

(sadly I had the midnight crash on my Boya system so only have recordings from mid - morning. I will post a full days comparison & multiday when available).

Early indications are that both systems have the same top 5 species (if you ignore the  oxpecker & bulbul).

Oh **AND** - the SEED hat has 

- a "BUTTON" (imagine what that can be programmed to do - like headless installation and setting in the middle of the Jungle outside of wifi access) and 
- additional plugs for GPIO and I2C interfacing, (temperature/humidity/light.......)

It would be great if others could try out the digital solution and help make it better (reduce noise/buzz, find optimal ALSA settings etc). If this is combined with the Analyze algorithm on a RPi Zero W.  A self contained standalone off the shelf system for around $30!! 

Just imagine.... one could get a 10 unit BirdNET-Pi-Weather Station array for the same price as whatever you want to spend $400 on.

Best
African Penguin

**Comment by SlavKarach** [comment]
This could be a very stupid question. I do not know much about audio at all. I barely even listen to music to be honest. However, what setup would I need for my RPI to achieve results similar to my iPhone app? Is your setup close to that? My phone detects more species more accurately at any time.

**Reply by CaiusX** [reply]
Hi Slav

The birdnet phone app doesn't work for me in South Africa, very few local species in the database. The BirdNet-Lite model has over 6,000 global species. 

The beauty of BirdNET-Pi is having an easily interrogated database and sound sample history.  It's an evolving system & the more users and suggestions/inputs from users and developers, the better it will get.

**Reply by SlavKarach** [reply]
Hi CaiusX,

I just checked my Merlin app (also made by the Cornell group) and looks like they have a group pack for SA. Did you try that one?

**Reply by CaiusX** [reply]
Yip, with poor results in the sound recognition section, photo id is good.

**Reply by ehpersonal38** [reply]
> This could be a very stupid question. I do not know much about audio at all. I barely even listen to music to be honest. However, what setup would I need for my RPI to achieve results similar to my iPhone app? Is your setup close to that? My phone detects more species more accurately at any time.

What's your minimum confidence? The app shows results even if there's a very low chance of it being right whereas BirdNET-Pi has a default of minimum 0.70 confidence.

**Comment by nilspupils** [comment]
Out of curiosity, i ordered the Boya and will give it a good test along with my other digital micrphone candidates soon. It would be great to find a solution which works without an interface card.
(Btw. my guess is that the Boya must have some kind of interface chip in the plug, but i have actually no idea how it works. I will investigate.) Perhaps someone else can answer.

**Reply by CaiusX** [reply]
Also give the Seed 2 Mic Hat a whirl if you can - quite inexpensive and has a button :)

**Comment by djdunc** [comment]
Hi @nilspupils am moving over here from this thread [in #39](https://github.com/mcguirepr89/BirdNET-Pi/discussions/39#discussioncomment-2721105 )  - on a new install RPi 4 (Bullseye Lite 64 bit) I followed the adafruit install guide plus the extra tweaks in alsamixer linked on that page to setup the software volume control - set it to about 58%. Initially it was at 3% so was not recording anything. With this setup I could then record sound, copy it across to a local machine and play it fine. So the mic works. Then installed BirdNet-Pi and based on tips in this thread tried various configurations of the audio card (default, 0, 1 - I noticed the card number sometimes changes on reboot..., even tried the card name) but the audio recordings are coming out silent. It "feels" like the correct audio input is not being set but I cannot figure out why.

**Reply by nilspupils** [reply]
@djdunc: Great you have it working so far! I think you are almost there with having the mic running! 
Can you get any other microphone to work (if you have one) with BirdNET-Pi right now? 
Also, can you still record audio with the I2S mic (is it still working outside of BirdNET but on the Pi)? 
You are officially the first one to try it out with a digital mic so we all have to learn it as we go along. Will try the same tomorrow with my system.

**Reply by mcguirepr89** [reply]
Hi, guys

I always recommend `default` because [device numbering isn't consistent](https://alsa-project.org/wiki/DeviceNames).

Take a look through your output for `arecord -L` to get the device _name_ to use for your "Audio Card" setting.

On my machine, for example:
```
birdnetpi1:~/BirdNET-Pi$ arecord -L
null
    Discard all samples (playback) or generate zero samples (capture)
default
    Playback/recording through the PulseAudio sound server
lavrate
    Rate Converter Plugin Using Libav/FFmpeg Library
samplerate
    Rate Converter Plugin Using Samplerate Library
speexrate
    Rate Converter Plugin Using Speex Resampler
jack
    JACK Audio Connection Kit
oss
    Open Sound System
pulse
    PulseAudio Sound Server
upmix
    Plugin for channel upmix (4,6,8)
vdownmix
    Plugin for channel downmix (stereo) with a simple spacialization
usbstream:CARD=Headphones
    bcm2835 Headphones
    USB Stream Output
usbstream:CARD=vc4hdmi0
    vc4-hdmi-0
    USB Stream Output
usbstream:CARD=vc4hdmi1
    vc4-hdmi-1
    USB Stream Output
hw:CARD=Device,DEV=0
    USB Advanced Audio Device, USB Audio
    Direct hardware device without any conversions
plughw:CARD=Device,DEV=0
    USB Advanced Audio Device, USB Audio
    Hardware device with all software conversions
sysdefault:CARD=Device
    USB Advanced Audio Device, USB Audio
    Default Audio Device
front:CARD=Device,DEV=0
    USB Advanced Audio Device, USB Audio
    Front output / input
dsnoop:CARD=Device,DEV=0
    USB Advanced Audio Device, USB Audio
    Direct sample snooping device
usbstream:CARD=Device
    USB Advanced Audio Device
    USB Stream Output
```

I would try the devices in this order:
1. `dsnoop:CARD=Device`
2. `usbstream:CARD=Device`
3. `hw:CARD=Device`

Upon each test, `journalctl -fu birdnet_recording.service -ulivestream.service` to see whether the device name you've chosen supports both services. If they fail because of "Number of channels not supported," try changing the number of channels from 2 to 1, or from 1 to 2.

I hope that is helpful!

Note: I just realized [(and fixed)](https://github.com/mcguirepr89/BirdNET-Pi/commit/ff8ceeb0f00dbdcc07f657d3df6cf7f23262e685) that the "Audio Card" variable needs quotes to not break the web interface, so if you do a quick update before any additional troubleshooting/testing, it will help a lot.

**Reply by nilspupils** [reply]
@mcguirepr89: Thanks for these hints! Didn't have time tonight but will try with a new installation tomorrow. I wonder if it makes a difference to install BirdNET-Pi first and then the digital microphone or the other way around. 
Apart from that, good to have you back!!!
Best, Nils

**Reply by nilspupils** [reply]
@djdunc: Please report back if you have still troubles after doing an update.

**Reply by djdunc** [reply]
thanks @mcguirepr89 the suggestion re journalctl has helped a lot - have edited the advanced.php file and tried a couple of settings and started to see the problem. I have stopped Birdnet services and have been exploring the settings the work for arecord in the terminal. Originally I was using:

`arecord -f S32_LE -c2 -r 48000 -t wav -D dmic_sv -V mono -v file.wav`

that produces a recording with the software volume set at 60% (via alsamixer). If I run:

`arecord -D plughw:0 -c1 -r 48000 -f S32_LE -t wav -V mono -v file.wav`

that produces a recording but is pretty much silent (audio doesn't go above 3%)

If I use default then I get no audio. 

I tried changing the format from S32_LE to S16_LE with the dmic_sv device but got an error saying `arecord: set_params:1343: Sample format non available`

arecord -l shows:

```
**** List of CAPTURE Hardware Devices ****
card 0: sndrpii2scard [snd_rpi_i2s_card], device 0: simple-card_codec_link snd-soc-dummy-dai-0 [simple-card_codec_link snd-soc-dummy-dai-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

arecord -L shows:
```
rpi-birdnet-i2c:~$ arecord -L
null
    Discard all samples (playback) or generate zero samples (capture)
default
    Playback/recording through the PulseAudio sound server
lavrate
    Rate Converter Plugin Using Libav/FFmpeg Library
samplerate
    Rate Converter Plugin Using Samplerate Library
speexrate
    Rate Converter Plugin Using Speex Resampler
jack
    JACK Audio Connection Kit
oss
    Open Sound System
pulse
    PulseAudio Sound Server
upmix
    Plugin for channel upmix (4,6,8)
vdownmix
    Plugin for channel downmix (stereo) with a simple spacialization
dmic_hw
dmic_sv
hw:CARD=sndrpii2scard,DEV=0
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Direct hardware device without any conversions
plughw:CARD=sndrpii2scard,DEV=0
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Hardware device with all software conversions
sysdefault:CARD=sndrpii2scard
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Default Audio Device
dsnoop:CARD=sndrpii2scard,DEV=0
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Direct sample snooping device
usbstream:CARD=sndrpii2scard
    snd_rpi_i2s_card
    USB Stream Output
usbstream:CARD=Headphones
    bcm2835 Headphones
    USB Stream Output
usbstream:CARD=vc4hdmi0
    vc4-hdmi-0
    USB Stream Output
usbstream:CARD=vc4hdmi1
    vc4-hdmi-1
    USB Stream Output
```

**Reply by djdunc** [reply]
for info am also copying in the full output from running arecord using dmic_sv in case that helps 

<img width="697" alt="Screenshot 2022-05-11 at 22 26 35" src="https://user-images.githubusercontent.com/145232/167950682-37d77e70-71f5-4654-aafb-c1fabc70f978.png">

**Reply by CaiusX** [reply]
Hi @djdunc 

If you're still struggling maybe try the KISS principle as partially outlined in the first post here  https://github.com/mcguirepr89/BirdNET-Pi/discussions/230#discussion-4044399

- arecord -l

- put the card # into settings rather than default (0 judging by your output, not sure if its "0" based on Patrick's comment above)- 
 
- run alsamixer in the Settings -> web terminal, 
 
- select the recording device (F6)
 
- set the recording volume (F4)
 

Best

**Reply by djdunc** [reply]
cheers @CaiusX - have stepped through those. I think it is down to some of the recording settings used in the arecord call being made by bird net - tonight I will dive into making some changes and see what I can find.

**Reply by mcguirepr89** [reply]
Hi, @djdunc 

From your output, it looks like you should try these, in this order:
1. `dsnoop:CARD=sndrpii2scard`
2. `usbstream:CARD=sndrpii2scard`
3. `hw:CARD=sndrpii2scard`

Please take another look at my example for why I've chosen those three.

I hope that helps!

**Reply by djdunc** [reply]
@mcguirepr89 thanks for help on this - finally have working version. To summarise for others (@nilspupils ) steps I had to go through were:
1. install mems mic according to Adafruit example https://learn.adafruit.com/adafruit-i2s-mems-microphone-breakout/raspberry-pi-wiring-test  (note I used BirdNet appropriate build for OS Lite 64 bit)
2. add the volume control to my alsa config by editing the .asoundrc file (at bottom of link in step 1) I set my volume in alsamixer to about 60%
3. in the Birdnet tools > settings > advanced settings I set the audio card value = dmic_sv
4. the last step was a hack - I had to edit the `BirdNET-Pi/scripts/birdnet_recording.sh` file to edit the `arecord -f S16_LE` setting to be S32_LE - ie:
```
    if [ -z ${REC_CARD} ];then
      arecord -f S32_LE -c${CHANNELS} -r48000 -t wav --max-file-time ${RECORDING_LENGTH}\
	      --use-strftime ${RECS_DIR}/%B-%Y/%d-%A/%F-birdnet-%H:%M:%S.wav
    else
      arecord -f S32_LE -c${CHANNELS} -r 48000 -t wav --max-file-time ${RECORDING_LENGTH}\
        -D "${REC_CARD}" --use-strftime \
	${RECS_DIR}/%B-%Y/%d-%A/%F-birdnet-%H:%M:%S.wav
    fi
```

Am now comparing the recordings against an Ultramic 192. The spectrograms look reasonably similar but the detections seem to be fewer - but will leave them running side by side for a day.

So, initial reaction is I will probably not use the MEMS and will look for a USB mic - but I will compare the results from a days listening tomorrow!

MEMS to left, Ultramic to right :

<img width="763" alt="Screenshot 2022-05-12 at 20 12 33" src="https://user-images.githubusercontent.com/145232/168152419-77e44d5a-062b-47aa-a686-56d01b4edaf1.png">

**Reply by nilspupils** [reply]
@djdunc: Awesome you got it working and thanks for the detailed instructions! The spectrogram looks really good, i wonder how it performs in the long run.

**Reply by djdunc** [reply]
Final update 24 hours later - both mics placed outside, pointing in more or less the same direction: 

I2S MEMS mic:
<img width="916" alt="Screenshot 2022-05-13 at 21 01 37" src="https://user-images.githubusercontent.com/145232/168381291-e1f47c3d-4db9-4ee7-99bd-c0278f333320.png">

UltraMic
<img width="824" alt="Screenshot 2022-05-13 at 21 01 53" src="https://user-images.githubusercontent.com/145232/168381334-32d8e7fb-723f-476c-a67c-960047ff8408.png">

MEMS was £6, Ultramic £200 - whilst the former was a little more fiddly, it performs pretty well. BTW also tested the AudioMoth in USB mode which worked really well - will be trying to get hold of some of their uMoths.

**Comment by srchild** [comment]
I can't make comparisons as it is all that I have tried, but as I had it already I tried using a Rode Smartlav+ microphone which I got a while ago on eBay for a good price and seems to work ok. (Smartlav+ has a higher specification than Smartlav). It can take the Rode foam or fluffy wind protection.

It has a TRRS plug so I connected it to ugreen interface [like this](https://smile.amazon.co.uk/UGREEN-Headphone-Microphone-Converter-Compatible/dp/B087T5H3MQ) (though it was cheaper on eBay).

**Reply by nilspupils** [reply]
@srchild: Thanks for the info and great you got it working with your setup. I think we will make a list of good/working microphones and soundcards to get these informations sorted in one place. Please post this also to here (for analog microphones and soundcard setups): 
https://github.com/mcguirepr89/BirdNET-Pi/discussions/39
